GlobeControls limit Frustum with horizon for performance

AlaricBaraou opened the pull request:
Currently for simplicity GlobeControls camera far is far enough to include the entire earth resulting in unnecessary tiles being loaded.

With these changes we limit the frustum to the distance of the horizon based on our current latitude and elevation.

This seems to work well but some tiles are being loaded while they don't intersect the frustum.

I edited the googleMapsExample.js to visualize it.

I'm not sure where is the issue but I'm pretty sure it's either the frustum or the OOB.
This particular example might help identifying the issue, when the camera is zoomed in on the north pole we get false positive on the intersectsFrustum check. 
If we look at it from the side, it's considered like an intersection as it seems like the frustum far limit does intersect with the obbs.
![wrongFrustumCheck](https://github.com/NASA-AMMOS/3DTilesRendererJS/assets/7174039/f309ce0f-8af7-4d28-a49d-6a841c0d1f68)
But if we look at it from above, we can see that the frustum doesn't.
![wrongFrustumCheckFromAbove](https://github.com/NASA-AMMOS/3DTilesRendererJS/assets/7174039/c47d7258-eca8-430f-a307-cb8c7ec635a5)

Those false positive unload as soon as on the side view, the frustum move outside of those obb.

gkjohnson commented:
Awesome! I'll take a deeper look at this next week. It would be nice to have some before / after stats on the number of tiles loaded and number of tiles rendered after the changes you've made so far.

> I'm not sure where is the issue but I'm pretty sure it's either the frustum or the OOB.
> This particular example might help identifying the issue, when the camera is zoomed in on the north pole we get false positive on the intersectsFrustum check.

I took a look and the issue is in the OBB frustum intersect code. In fact three.js' AABB frustum intersect code that the OBB logic is based on isn't correct, either. It will mark bounds as in the frustum even when they're clearly outside of it since it's only checking if the box is completely on one side of any plane which misses the common case of a box that intersects multiple planes if it's large. Here's a demo:

<img width="226" alt="image" src="https://github.com/NASA-AMMOS/3DTilesRendererJS/assets/734200/bb5e3474-2d71-4c77-ba81-7fc6e23f0648">

https://jsfiddle.net/89rtuaqp/2/

[This post](https://iquilezles.org/articles/frustumcorrect/) discusses the problem and a correct application of separating axis theorem. I'll make a three.js issue. We can see what they say - we may have to make our own extended frustum class that stores enough data for performing a more correct OBB intersection test.

gkjohnson commented:
https://github.com/mrdoob/three.js/issues/27756 may take a bit to resolve so I think it's best for us to modify our current implementation to be more robust, but I think that can happen in another PR. This can be done by storing the frustum corners as points on the Frustum so they can be tested against the OBB planes as referenced in the inigo post.

AlaricBaraou commented:
Good timing, I was working on a custom Frustum but it's hard to test it with the tiles logic, I might try in a simpler environment.

I added a class AdvFrustum that calculate the points of the Frustum to use them for different operations as suggested in the article you linked. But I think there are still case that are not handled or it might be the logic of the nested tiles etc that I don't really grasp at the moment.

Also the latest commit take care of "Google Maps Example: Add ability to specify height in hash" #482

AlaricBaraou commented:
a simpler example to play with it 

https://jsfiddle.net/0sb3gden/30/

The extended Frustum works well with axis aligned box but not with OBB at the moment.

gkjohnson commented on src/three/math/OBB.js:
```diff
@@ -13,9 +14,10 @@ export class OBB {
 	update() {
 
-		const { points, inverseTransform, transform, box } = this;
+		const { points, inverseTransform, transform, box, orientedBox } = this;
 		inverseTransform.copy( transform ).invert();
 
+		orientedBox.copy( box ).applyMatrix4( transform );
```
If you transform the box by the transform you just get an enlarged AABB. We'll want to store the planes of the OBB (or generate them on the fly) so we can use those to test the frustum points.

gkjohnson commented on src/three/math/OBB.js:
```diff
@@ -64,6 +66,13 @@ export class OBB {
 
 		}
 
+		const boxInFrustum = frustum.boxInFrustum( this.orientedBox );
+		if ( ! boxInFrustum ) {
+
+			return false;
+
+		}
+
```
Just like we had to make a custom implementation for testing all the OBB points against the frustum planes above -we'll need to do the same thing for the frustum points against the OBB planes here. The "boxInFrustum" function shouldn't be needed in this case.

AlaricBaraou commented:
I was about to write that it's good now but apparently now we have some false negative.

Looking good here
![frustumOk](https://github.com/NASA-AMMOS/3DTilesRendererJS/assets/7174039/3cfccb54-a869-4cbe-88ca-15cb5db3ddb4)

But here it discard some tiles that are in obvious intersections. 
![frustumNOk](https://github.com/NASA-AMMOS/3DTilesRendererJS/assets/7174039/636b02c9-9e88-40cd-a1b8-380f2b629b1f)

I`ll try to reproduce that one in jsfiddle

gkjohnson commented on src/three/math/Ellipsoid.js:
```diff
@@ -200,4 +203,43 @@ export class Ellipsoid {
+	}
+
+	getPositionElevation( pos ) {
+
+		_pos.copy( pos );
+		this.worldToLocal( _pos );
+
+		this.getPositionToSurfacePoint( _pos, _vec3 );
+		const elevation = _vec3.distanceTo( _pos );
+
+		return elevation;
```
There shouldn't be any notion of a transform here - the 3d point should be passed in the same frame that the ellipsoid is stored in. If we assume a 90 degree rotation here then if the rotation of the globe object is set to anything different then this will break.

The 90 globe rotation is set on the tile root group [here](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/5d156142485764e1c1d64d3d024f8e7a9efb6edb/example/googleMapsExample.js#L52). So calling this function should look like this:

```js
_invMatrix.copy( tilesRenderer.group.matrixWorld ).invert();
_tempVec.copy( camera.position ).applyMatrix4( _invMatrix );
const elevation = ellipsoid.getPositionElevation();
```

The same pattern is used in GlobeControls [here](https://github.com/NASA-AMMOS/3DTilesRendererJS/blob/5d156142485764e1c1d64d3d024f8e7a9efb6edb/src/three/controls/GlobeControls.js#L101-L103).

gkjohnson commented on src/three/math/Ellipsoid.js:
```diff
@@ -200,4 +203,43 @@ export class Ellipsoid {
 
 	}
 
+	calculateHorizonDistance( latitude, elevation ) {
+
+		// elevation relative to the surface of the ellipsoid at the specified latitude
+		const effectiveRadius = this.calculateEffectiveRadius( latitude );
+		return Math.sqrt( 2 * effectiveRadius * elevation + elevation ** 2 );
+
+	}
+
+	calculateEffectiveRadius( latitude ) {
+
+		const semiMajorAxis = this.radius.x;
+		const semiMinorAxis = this.radius.z;
+		const eSquared = 1 - ( semiMinorAxis ** 2 / semiMajorAxis ** 2 );
+		const phi = latitude * MathUtils.DEG2RAD;
+
+		const sinPhiSquared = Math.sin( phi ) ** 2;
+		const N = semiMajorAxis / Math.sqrt( 1 - eSquared * sinPhiSquared );
+		return N;
+
+	}
```
Are these formulas from somewhere? Can we add more comments?

AlaricBaraou replied:
I added source for the formulas

gkjohnson commented on src/three/math/OBB.js:
```diff
@@ -17,25 +24,42 @@ export class OBB {
+	updatePlanes() {
 
-				}
+		if ( ! this.planeNeedsUpdate ) return;
 
-			}
+		const normals = [
+			_normal1.subVectors( this.points[ 1 ], this.points[ 0 ] ).cross( _vec3.subVectors( this.points[ 2 ], this.points[ 0 ] ) ).normalize(),
+			_normal2.subVectors( this.points[ 4 ], this.points[ 0 ] ).cross( _vec3.subVectors( this.points[ 1 ], this.points[ 0 ] ) ).normalize(),
+			_normal3.subVectors( this.points[ 2 ], this.points[ 0 ] ).cross( _vec3.subVectors( this.points[ 4 ], this.points[ 0 ] ) ).normalize()
+		];
 
-		}
+		// Front and back
+		this.planes[ 0 ].setFromNormalAndCoplanarPoint( normals[ 0 ], this.points[ 0 ] );
+		this.planes[ 1 ].setFromNormalAndCoplanarPoint( normals[ 0 ].negate(), this.points[ 6 ] );
+
+		// Left and right
+		this.planes[ 2 ].setFromNormalAndCoplanarPoint( normals[ 1 ], this.points[ 0 ] );
+		this.planes[ 3 ].setFromNormalAndCoplanarPoint( normals[ 1 ].negate(), this.points[ 2 ] );
+
+		// Top and bottom
+		this.planes[ 4 ].setFromNormalAndCoplanarPoint( normals[ 2 ], this.points[ 0 ] );
+		this.planes[ 5 ].setFromNormalAndCoplanarPoint( normals[ 2 ].negate(), this.points[ 1 ] );
+
+		this.planeNeedsUpdate = false;
```
I think these functions have become a bit difficult to follow. Instead of pulling points out of the points array we can generate the planes in a more explicit method that's hopefully easier to validate:

```js
_worldMin.copy( this.box.min ).applyMatrix4( this.transform );
_worldMax.copy( this.box.max ).applyMatrix4( this.transform );

_norm.set( 0, 0, 1 ).transformDirection( this.transform );
plane[ 0 ]. setFromNormalAndCoplanarPoint( _norm, _worldMin );
plane[ 1 ]. setFromNormalAndCoplanarPoint( _norm, _worldMax ).negate();

// ...
```

Note that just like the frustum planes the "positive" side of the plane will need to face "inward".

AlaricBaraou replied:
oh I missed this one, I'll use setFromNormalAndCoplanarPoint instead

gkjohnson commented on src/three/math/OBB.js:
```diff
@@ -17,25 +24,42 @@ export class OBB {
+		points[ 5 ].set( max.x, min.y, max.z ).applyMatrix4( transform ); // Back-bottom-right
+		points[ 6 ].set( max.x, max.y, max.z ).applyMatrix4( transform ); // Back-top-right
+		points[ 7 ].set( min.x, max.y, max.z ).applyMatrix4( transform ); // Back-top-left
+		this.planeNeedsUpdate = true;
```
I don't think there's any reason to defer this plane generation, right? We can just update them here after the points in the "update" function.

AlaricBaraou replied:
The idea was often / sometimes, the planes are not needed as intersectsFrustum can return early before reaching the planes check.
I didn't test for any tradeoff or if neglectable. 
I'll go with the idea you prefer.

gkjohnson reviewed (commented):
It looks like we're getting some depth fighting issues - but it looks like this was already happening 🤔 

<img width="1154" alt="image" src="https://github.com/NASA-AMMOS/3DTilesRendererJS/assets/734200/d2cd5e50-7a3c-43f5-be7e-1ea9f624e845">

We can look at this in another PR but we might want to find a smarter way to set the "near" value.

gkjohnson commented on src/three/math/AdvFrustum.js:
```diff
@@ -0,0 +1,64 @@
+		// Solve for X by applying the inverse matrix to B
+		const X = B.applyMatrix3( A.invert() );
+
+		return [ X.x, X.y, X.z ];
```
Let's use the common three.js pattern of passing a `target` vector variable into this function so we can avoid allocating a new array.

gkjohnson commented on src/three/math/AdvFrustum.js:
```diff
@@ -0,0 +1,64 @@
+const _mat3 = new Matrix3();
+const _constants = new Vector3();
+
+class AdvFrustum extends Frustum {
```
_Tiny nit_ if you're in here modifying stuff - I'd call this "ExtendedFrustum".

gkjohnson commented on src/three/controls/GlobeControls.js:
```diff
@@ -152,7 +153,19 @@ export class GlobeControls extends EnvironmentControls {
-		camera.far = distanceToCenter + largestDistance + 0.1;
+
+		// update the far plane to the horizon distance
+		ellipsoid.getPositionToCartographic( camera.position, _vec );
```
This function call doesn't look right? The camera position value needs to be transformed into the local "tilesGroup" space, first. This function also should take an empty object since it assigns values to "lat" and "lon". Right now this is modifying a "Vector3" object.

AlaricBaraou replied:
yes nice catch, I guess the difference / influence of that variable was minimal I don't notice any change, but yes it's absolutely the right way to do it.

gkjohnson commented on example/googleMapsExample.js:
```diff
@@ -190,6 +238,17 @@ function animate() {
 	renderer.render( scene, camera );
 	stats.update();
 
+	// render third person view
+	thirdPersonRenderer.domElement.style.visibility = params.showThirdPerson ? 'visible' : 'hidden';
+	if ( params.showThirdPerson ) {
+
+		cameraHelper.update();
+		cameraHelper.visible = true;
+
+		thirdPersonRenderer.render( scene, thirdPersonCamera );
+
+	}
+
```
Do we need to keep all of these debug rendering additions? Ideally this example would just show a simple approach to adding a new google maps globe without too much extra.

AlaricBaraou replied:
no need indeed, it was just for me

gkjohnson reviewed (commented):
Added some comments for some small things to change - and a few questions on the example and GlobeControls additions.

AlaricBaraou commented on test/Ellipsoid.test.js:
```diff
@@ -101,7 +101,7 @@ describe( 'Ellipsoid', () => {
 		for ( let i = 0; i < 100; i ++ ) {
 
-			v.random().normalize().multiplyScalar( WGS84_RADIUS * 2 );
+			v.random().normalize().multiplyScalar( WGS84_HEIGHT * 2 );
```
I was making tests and forgot to replace with the correct variable. fixing it

AlaricBaraou commented on src/three/math/OBB.js:
```diff
@@ -17,25 +23,39 @@ export class OBB {
 		inverseTransform.copy( transform ).invert();
 
 		const { min, max } = box;
-		let index = 0;
-		for ( let x = - 1; x <= 1; x += 2 ) {
+		points[ 0 ].set( min.x, min.y, min.z ).applyMatrix4( transform ); // Front-bottom-left
+		points[ 1 ].set( max.x, min.y, min.z ).applyMatrix4( transform ); // Front-bottom-right
+		points[ 2 ].set( max.x, max.y, min.z ).applyMatrix4( transform ); // Front-top-right
+		points[ 3 ].set( min.x, max.y, min.z ).applyMatrix4( transform ); // Front-top-left
 
-			for ( let y = - 1; y <= 1; y += 2 ) {
+		points[ 4 ].set( min.x, min.y, max.z ).applyMatrix4( transform ); // Back-bottom-left
+		points[ 5 ].set( max.x, min.y, max.z ).applyMatrix4( transform ); // Back-bottom-right
+		points[ 6 ].set( max.x, max.y, max.z ).applyMatrix4( transform ); // Back-top-right
+		points[ 7 ].set( min.x, max.y, max.z ).applyMatrix4( transform ); // Back-top-left
```
Since the update to updatePlanes yes, doing it now

AlaricBaraou commented on src/three/math/AdvFrustum.js:
```diff
@@ -0,0 +1,64 @@
+
+	}
+
+	findIntersectionPointThree( plane1, plane2, plane3 ) {
+
+		// Create the matrix A using the normals of the planes
+		const A = _mat3.set(
+			plane1.normal.x, plane1.normal.y, plane1.normal.z,
+			plane2.normal.x, plane2.normal.y, plane2.normal.z,
+			plane3.normal.x, plane3.normal.y, plane3.normal.z
+		);
+
+		// Create the vector B using the constants of the planes
+		const B = _constants.set( - plane1.constant, - plane2.constant, - plane3.constant );
+
+		// Solve for X by applying the inverse matrix to B
+		const X = B.applyMatrix3( A.invert() );
+
+		return [ X.x, X.y, X.z ];
+
+	}
```
Make sense, I'll try to follow that one from now on, updating it

AlaricBaraou commented:
I addressed all points, just a note about this issue https://github.com/NASA-AMMOS/3DTilesRendererJS/issues/482 
It can't be init under a certain height, at least in the area I was testing it.
I added a comment to check later

gkjohnson commented:
> It can't be init under a certain height, at least in the area I was testing it.

Likely what's happening is that the coarse levels of detail on the model load and have a much higher height than the lower LoDs which pushes the camera up due to the "GlobeControls" update function. I think it's okay to say that this fixes #482.

gkjohnson commented:
A few stats - some very nice improvements!

**Top Down View**

| | Before | After | % improvement |
|---|---|---|---|
| visible | 104 | 55 | ~47% |
| loaded | 307 | 189 | ~38% |

<img width="350" alt="image" src="https://github.com/NASA-AMMOS/3DTilesRendererJS/assets/734200/6df34887-fd61-4cfd-b772-7b2139838161">

**Horizon View**

| | Before | After | % improvement |
|---|---|---|---|
| visible | 332 | 186 | ~43% |
| loaded | 574 | 376 | ~34% |

<img width="350" alt="image" src="https://github.com/NASA-AMMOS/3DTilesRendererJS/assets/734200/ba7e3c47-4dd2-414e-acd2-2f5400c9df0c">

gkjohnson commented:
One thing I did notice is that the globe controls far plane _will_ clip the horizon a bit. It's subtle but you can see it change a little. Likely due to the earth model not exactly matching the ellipsoid. We don't have to fix it immediately but may something to keep in mind for the future there are ideas on how to address it.

**Before**

<img width="511" alt="image" src="https://github.com/NASA-AMMOS/3DTilesRendererJS/assets/734200/027a6bb5-16b9-4e88-a1c2-c7a0c45d93b7">

**After**
<img width="460" alt="image" src="https://github.com/NASA-AMMOS/3DTilesRendererJS/assets/734200/5176d9e6-62a2-4e34-b560-85358f220e7d">
